source/request.test.js

```diff
@@ -0,0 +1,212 @@
+const Request = require('./request');
+
+test('preprocesses headers', () => {
+  let details = {
+    responseHeaders: [
+      {
+        name: 'server',
+        value: 'cloudflare'
+      }
+    ]
+  };
+
+  let r = new Request(details);
+
+  expect(r.headers).toMatchObject({
+    SERVER: 'cloudflare'
+  });
+});
+
+describe('Railgun', () => {
+  test('processes Railgun header v1', () => {
+    let details = {
+      responseHeaders: [
+        {
+          name: 'cf-railgun',
+          value: 'normal 96 foobar 9001'
+        }
+      ]
+    };
+
+    let r = new Request(details);
+
+    expect(r.railgunMetaData).toMatchObject({
+      normal: true,
+      id: 'normal',
+      version: '9001',
+      flags: 0x60,
+      messages: [
+        'Dictionary found in memcache',
+        'Restarted broken origin connection'
+      ]
+    });
+  });
+
+  test('processes Railgun header v2', () => {
+    let details = {
+      responseHeaders: [
+        {
+          name: 'cf-railgun',
+          value: 'foobar 46 700 96 9001'
+        }
+      ]
+    };
+
+    let r = new Request(details);
+
+    expect(r.railgunMetaData).toMatchObject({
+      normal: false,
+      id: 'foobar',
+      version: '9001',
+      compression: '54%',
+      time: '700sec',
+      flags: 0x60,
+      messages: [
+        'Dictionary found in memcache',
+        'Restarted broken origin connection'
+      ]
+    });
+  });
+
+  test('determines Railgun', () => {
+    let detailsRG = {
+      responseHeaders: [
+        {
+          name: 'cf-railgun',
+          value: 'normal'
```

Comment:
Think this test should be expanded to cover the case where the request could've been served by Railgun even in the absence of the `'normal'` string the header.

Also, I forgot what this normal business is, can't remember if the `cf-railgun` header was there when I originally wrote this.

Would recommend using (close to) real header values here if possible and testing the _normal_ style, v2 and non-railgun case. Go has a good pattern for tests like this one - [table driven tests](https://dave.cheney.net/2013/06/09/writing-table-driven-tests-in-go)

Reply:
I think v2 is the only possible output now, but I haven't confirmed.

I considered a table based test, but wanted to get the basics in first.

Reply:
👍